addons/website_slides_survey/models/survey_user.py

```diff
@@ -11,3 +12,61 @@ class SurveyUserInput(models.Model):
         help="The related course slide when there is no membership information")
     slide_partner_id = fields.Many2one('slide.slide.partner', 'Subscriber information',
         help="Slide membership information for the logged in user")
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        records = super(SurveyUserInput, self).create(vals_list)
+        records._check_for_failed_attempt()
+        return records
+
+    @api.multi
+    def write(self, vals):
+        res = super(SurveyUserInput, self).write(vals)
+        if 'state' in vals:
+            self._check_for_failed_attempt()
+        return res
+
+    def _check_for_failed_attempt(self):
+        """ If the user fails his last attempt at a course certification,
+        we remove him from the members of the course (and he has to enroll again).
+        He receives an email in the process notifying him of his failure and suggesting
+        he enrolls to the course again.
+
+        The purpose is to have a 'certification flow' where the user can re-purchase the
+        certification when they have failed it."""
+
+        if self:
+            user_inputs = self.search([
+                ('id', 'in', self.ids),
+                ('state', '=', 'done'),
+                ('quizz_passed', '=', False),
+                ('slide_partner_id', '!=', False)
+            ])
+
+            if user_inputs:
+                removed_slide_partner_domain = []
+                removed_channel_partner_domain = []
+                for user_input in user_inputs:
+                    if user_input.survey_id._has_attempts_left(user_input.partner_id, user_input.email, user_input.invite_token):
+                        # skip if user still has attempts left
+                        continue
+
+                    self.env.ref('website_slides_survey.mail_template_user_input_certification_failed').send_mail(
+                        user_input.id, notif_layout="mail.mail_notification_light"
+                    )
+                    slide_partner_id = user_input.slide_partner_id
+                    removed_slide_partner_domain = expression.OR([
```

Comment:
Actually why not removed_slide_partners |= slide_partner_id ?

Reply:
Because when you fail a single slide (certification) of the channel we need te remove all slide_partner relations inside the channel, not just from the slide you failed.

Does that make sense?

Reply:
Oh, indeed. Maybe we should move this to a tool method _remove_members on channel model.

Reply:
I created the tool method as suggested and the code looks cleaner now.

Let me know if it looks like what you had in mind.